moshi/src/main/java/com/squareup/moshi/ClassFactory.kt

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2011 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import com.squareup.moshi.internal.Util
+import java.io.ObjectInputStream
+import java.io.ObjectStreamClass
+import java.lang.reflect.InvocationTargetException
+import kotlin.Throws
+
+/**
+ * Magic that creates instances of arbitrary concrete classes. Derived from Gson's UnsafeAllocator
+ * and ConstructorConstructor classes.
+ *
+ * @author Joel Leitch
+ * @author Jesse Wilson
+ */
+internal abstract class ClassFactory<T> {
+  @Throws(InvocationTargetException::class, IllegalAccessException::class, InstantiationException::class)
+  abstract fun newInstance(): T
+
+  companion object {
+    @JvmStatic
+    operator fun <T> get(rawType: Class<*>): ClassFactory<T> {
+      // Try to find a no-args constructor. May be any visibility including private.
+      try {
+        val constructor = rawType.getDeclaredConstructor()
+        constructor.isAccessible = true
+        return object : ClassFactory<T>() {
+          @Suppress("UNCHECKED_CAST")
+          override fun newInstance() = constructor.newInstance() as T
+
+          override fun toString() = rawType.name
+        }
+      } catch (ignored: NoSuchMethodException) {
+        // No no-args constructor. Fall back to something more magical...
+      }
+
+      // Try the JVM's Unsafe mechanism.
+      // public class Unsafe {
+      //   public Object allocateInstance(Class<?> type);
+      // }
+      try {
+        val unsafeClass = Class.forName("sun.misc.Unsafe")
+        val f = unsafeClass.getDeclaredField("theUnsafe")
```

Comment:
Let's call this `unsafeField` while we're at it

Reply:
https://github.com/square/moshi/pull/1469/commits/de0bd183df3dc489aae06577767cdcd7bda06be4#diff-7138ff914c5ee796a257ccefff821c0f10e5465884572131cade205a3cd24256R58